File: lumen-examples/rev_number/seq2seq.rs
use anyhow::Context;
use lumen_core::{FloatDType, IndexOp, IntTensor, NoGradGuard, Tensor};
use lumen_dataset::{Batcher, DataLoader, Dataset};
use lumen_nn::{functional::LossReduction, optim::{Adam, AdamConfig, Optimizer}, CrossEntropyLoss, Embedding, Linear, Lstm, LstmState, Module};
use rand::{rng, Rng};

fn main() {
    if let Err(e) = result_main() {
        eprintln!("Err: {:?}", e);
    }
}

#[derive(Debug, Clone)]
pub struct TokenConfig {
    pub pad_token: u32,
    pub min_token: u32,
    pub max_token: u32,
    pub sos_token: u32,
    pub eos_token: u32,
    pub vocab_size: usize,
}

impl TokenConfig {
    pub const fn new(min_token: u32, max_token: u32) -> Self {
        assert!(max_token > min_token, "max token <= min token");
        Self {
            min_token, max_token,
            pad_token: max_token + 1,
            sos_token: max_token + 2,
            eos_token: max_token + 3,
            vocab_size: (max_token + 3 - min_token + 1) as usize,
        }
    }
}
 
fn result_main() -> anyhow::Result<()> {
    const TOKEN_CONFIG: TokenConfig = TokenConfig::new(0, 10);

    const EMBED_SIZE: usize = 16;
    const HIDDEN_SIZE: usize = 32;
    const MAX_SEQ_LEN: usize = 8;
    const MIN_SEQ_LEN: usize = 4;
    const LR: f32 = 0.01;
    const EPOCHS: usize = 20;
    const DATASET_LEN: usize = 1000;
    const BATCH_SIZE: usize = 10;

    let dataset = Seq2SeqDataset::new(TOKEN_CONFIG, MAX_SEQ_LEN, MIN_SEQ_LEN, DATASET_LEN);
    let dataloader = Seq2SeqDataLoader::new(dataset, Seq2SeqBatcher::new(TOKEN_CONFIG), BATCH_SIZE, true);

    let mut model = Seq2Seq::<f32>::new(TOKEN_CONFIG, EMBED_SIZE, HIDDEN_SIZE).context("new model")?;
    let mut config = AdamConfig::default();
    config.lr = LR;
    let mut optimizer = Adam::new(model.params(), config)?;
    let criterion = CrossEntropyLoss::new(LossReduction::Mean);

    for epoch in 0..EPOCHS {
        for (batch_idx, batch) in dataloader.iter().enumerate() {
            // (batch_size, seq_len), (batch_size, seq_len+1), 
            let (src, trg) = batch.context("get batch")?;

            // (batch, seq_len+1, vocab_size)
            let output = model.forward(&src, Some(&trg)).context("model forward")?;

            // (batch, seq_len+1, vocab_size) => (batch * (seq_len+1), vocab_size)
            let output = output.flatten(0, 1)?;
            // (batch_size * (seq_len+1), 1)
            let trg = trg.flatten_all()?.unsqueeze(1)?;
            let loss = criterion.forward(&output, &trg).context("loss")?;
            
            let grads = loss.backward().context("backward")?;

            optimizer.step(&grads).context("step")?;

            if batch_idx % 10 == 0 {
                println!("epoch {}, batch {}, loss = {}", epoch, batch_idx, loss.to_scalar()?);
            }
        }
    }

    {
        let _guard = NoGradGuard::new();
        model.eval();

        let test_dataset = Seq2SeqDataset::new(TOKEN_CONFIG, MAX_SEQ_LEN, MIN_SEQ_LEN, 5);
        let test_loader = Seq2SeqDataLoader::new(test_dataset, Seq2SeqBatcher::new(TOKEN_CONFIG), 1, true);

        for (i, batch) in test_loader.iter().enumerate() {
            // (1, seq_len), (1, seq_len+1)
            let (src, trg) = batch.context("batch")?;
            // (1, seq_len) => (1, seq_len+1)
            let seqs = model.prediect(&src).context("prediect")?;

            println!("Batch {}:", i);
            println!("Pred: {:?}", seqs[0]);
            println!("Anwe: {:?}", &trg.flatten_all()?.to_vec()?);
        }
    }

    Ok(())
}

#[derive(Module)]
pub struct Encoder<T: FloatDType> {
    pub embedding: Embedding<T>,
    pub rnn: Lstm<T>,
}

impl<T: FloatDType> Encoder<T> {
    pub fn new(input_size: usize, embed_size: usize, hidden_size: usize) -> anyhow::Result<Self> {
        let embedding = Embedding::new(input_size, embed_size, None)?;
        let rnn = Lstm::new(embed_size, hidden_size, None)?;
        Ok(Self { embedding, rnn })
    }

    /// src: (batch_size, seq_len)
    pub fn forward(&self, src: impl Into<IntTensor>) -> anyhow::Result<LstmState<T>> {
        // (batch_size, seq_len) => (batch_size, seq_len, embed_size)
        let embedded = self.embedding.forward(src).context("embedding forward")?;
        let (_, state) = self.rnn.forward(&embedded, None).context("rnn forward")?;
        Ok(state)
    } 
}

#[derive(Module)]
pub struct Decoder<T: FloatDType> {
    pub embedding: Embedding<T>,
    pub rnn: Lstm<T>,
    pub fc_out: Linear<T>,
}

impl<T: FloatDType> Decoder<T> {
    pub fn new(output_size: usize, embed_size: usize, hidden_size: usize) -> anyhow::Result<Self> {
        let embedding = Embedding::new(output_size, embed_size, None)?;
        let rnn = Lstm::new(embed_size, hidden_size, None)?;
        let fc_out = Linear::new(hidden_size, output_size, true, None)?;
        Ok(Self { embedding, rnn, fc_out })
    }

    /// Decoder forward
    /// 
    /// ## Args
    /// - input_token: (batch_size, 1)
    /// - state: from encoder / decoder
    pub fn forward(&self, input_token: impl Into<IntTensor>, state: &LstmState<T>) -> anyhow::Result<(Tensor<T>, LstmState<T>)> {
        // (batch_size, 1) => (batch_size, 1, embed_dim)
        let embedded = self.embedding.forward(input_token).context("embedding forward")?;
        // (batch_size, 1, hidden_size)
        let (output, state) = self.rnn.forward(&embedded, Some(state)).context("rnn forward")?;
        // (batch_size, 1, hidden_size) => (batch_size, 1, vocab_size)
        let prediction = self.fc_out.forward(&output).context("linear forward")?;

        Ok((prediction, state))
    }
}

#[derive(Module)]
pub struct Seq2Seq<T: FloatDType> {
    pub encoder: Encoder<T>,
    pub decoder: Decoder<T>,

    #[module(skip)]
    pub token_config: TokenConfig,
}

impl<T: FloatDType> Seq2Seq<T> {
    pub fn new(token_config: TokenConfig, embed_size: usize, hidden_size: usize) -> anyhow::Result<Self> {
        let encoder = Encoder::new(token_config.vocab_size, embed_size, hidden_size)?;
        let decoder = Decoder::new(token_config.vocab_size, embed_size, hidden_size)?;
        Ok(Self { token_config, encoder, decoder })
    }

    /// Forward pass
    /// 
    /// - src: (batch, seq_len)
    /// - trg: Option<&Tensor> -> (batch, seq_len + 1) [t1, t2, ..., EOS]
    pub fn forward(&self, src: &Tensor<u32>, trg: Option<&Tensor<u32>>) -> anyhow::Result<Tensor<T>> {
        /*

            src:
            +---+---+---+---+
            | 7 | 4 | 2 | P |
            +---+---+---+---+
            | 1 | 2 | 3 | 4 |
            +---+---+---+---+

            trg:
            +---+---+---+---+---+
            | 2 | 4 | 7 | E | P |
            +---+---+---+---+---+
            | 4 | 3 | 2 | 1 | E |
            +---+---+---+---+---+

        */
        let (batch, seq_len) = src.dims2()?;
        let mut outputs = vec![];

        // encoder
        let mut state = self.encoder.forward(src).context("encoder")?;
        let mut input_token = Tensor::new(self.token_config.sos_token)?.reshape((1, 1))?.repeat((batch, 1))?;
        
        for t in 0..seq_len + 1 {
            // (batch, 1, vocab_size)
            let (output, next_state) = self.decoder.forward(&input_token, &state).context("decoder")?;

            // update
            state = next_state;

            input_token = match trg {
                Some(trg) => trg.index((.., t))?.unsqueeze(1)?,
                None => output.argmax(2)?,    
            };

            // save
            outputs.push(output);
        }

        // [(batch_size, 1, vocab_size)] => (batch_size, seq_len+1, vocab_size)
        Tensor::cat(&outputs, 1).context("stack output")
    }

    /// src: (batch, seq_len)
    pub fn prediect(&self, src: &Tensor<u32>) -> anyhow::Result<Vec<Vec<u32>>> {
        // (batch_size, seq_len+1, vocab_size)
        let outputs = self.forward(&src, None).context("model forward")?;
        // (batch_size, seq_len+1)
        let pred_tokens = outputs.argmax(2).context("argmax output")?;

        let mut seqs = vec![];
        let batch = pred_tokens.dim(0)?;
        for b in 0..batch {
            let tokens = pred_tokens.index(b)?.to_vec()?;
            seqs.push(tokens);
        }

        Ok(seqs)
    }
}

pub struct Seq2SeqDataset {
    pub token_config: TokenConfig,
    pub max_seq_len: usize,
    pub min_seq_len: usize,
    pub dataset_len: usize,
}

impl Seq2SeqDataset {
    pub fn new(token_config: TokenConfig, max_seq_len: usize, min_seq_len: usize, dataset_len: usize) -> Self {
        assert!(max_seq_len >= min_seq_len, "max_seq_len < min_seq_len");
        Self { token_config, max_seq_len, min_seq_len,  dataset_len }
    }
}

impl Dataset for Seq2SeqDataset {
    type Item = (Vec<u32>, Vec<u32>);
    
    fn get(&self, _index: usize) -> Option<Self::Item> {
        let mut rng = rng();
        
        // rand a seq_len
        let seq_len = rng.random_range(self.min_seq_len..=self.max_seq_len);

        // generate input tokens
        let src: Vec<u32> = (0..seq_len).map(|_| rng.random_range(self.token_config.min_token..=self.token_config.max_token)).collect();
        
        // revser input and add eos
        let mut trg = src.clone();
        trg.reverse();
        trg.push(self.token_config.eos_token);

        Some((src, trg))
    }

    fn len(&self) -> usize {
        self.dataset_len
    }
}

pub struct Seq2SeqBatcher {
    pub token_config: TokenConfig,
}

impl Seq2SeqBatcher {
    pub fn new(token_config: TokenConfig) -> Self {
        Self { token_config }
    }
}

impl Batcher for Seq2SeqBatcher {
    type Item = (Vec<u32>, Vec<u32>);
    type Output = Result<(Tensor<u32>, Tensor<u32>), lumen_core::Error>;

    fn batch(&self, items: Vec<(Vec<u32>, Vec<u32>)>) -> Self::Output {
        let mut xs = vec![];
        let mut ys = vec![];
        
        // find max seq len!
        let mut max_seq_len = 0;
        for (x, y) in items.iter() {
            assert_eq!(x.len() + 1, y.len());
            if x.len() > max_seq_len {
                max_seq_len = x.len();
            }
        }

        for (mut x, mut y) in items {
            // add pad!
            for _ in x.len()..max_seq_len {
                x.push(self.token_config.pad_token);
            }

            for _ in y.len()..max_seq_len+1 {
                y.push(self.token_config.pad_token);
            }

            assert_eq!(x.len(), max_seq_len);
            assert_eq!(y.len(), max_seq_len + 1);

            xs.push(Tensor::new(x)?);
            ys.push(Tensor::new(y)?);
        }

        let xs = Tensor::stack(&xs, 0)?;
        let ys = Tensor::stack(&ys, 0)?; 
        Ok((xs, ys))
    }
}

pub type Seq2SeqDataLoader = DataLoader<Seq2SeqDataset, Seq2SeqBatcher>;

File: examples/rev_number/seq2seq.py
import torch
import torch.nn as nn
import torch.optim as optim
import random

# --- 1. 配置参数 ---
# 词表大小：数字 0-9，加上 SOS(开始标志)=10, EOS(结束标志)=11
VOCAB_SIZE = 12 
SOS_TOKEN = 10
EOS_TOKEN = 11

# 模型参数 (设得很小，CPU 随便跑)
EMBED_DIM = 16
HIDDEN_DIM = 32
SEQ_LEN = 6      # 序列长度
LR = 0.01
EPOCHS = 2000    # 迭代次数

# --- 2. 定义 Seq2Seq 模型 ---

class Encoder(nn.Module):
    def __init__(self, input_dim, emb_dim, hidden_dim):
        super(Encoder, self).__init__()
        # 把数字索引转换成向量
        self.embedding = nn.Embedding(input_dim, emb_dim)
        # LSTM 处理序列
        self.rnn = nn.LSTM(emb_dim, hidden_dim, batch_first=True)

    def forward(self, src):
        # src: [batch_size, seq_len]
        embedded = self.embedding(src) # -> [batch, seq_len, emb_dim]
        
        # LSTM 返回: output, (hidden, cell)
        # 我们只需要 hidden 和 cell (这就是 Context Vector)
        _, (hidden, cell) = self.rnn(embedded)
        
        return hidden, cell

class Decoder(nn.Module):
    def __init__(self, output_dim, emb_dim, hidden_dim):
        super(Decoder, self).__init__()
        self.embedding = nn.Embedding(output_dim, emb_dim)
        self.rnn = nn.LSTM(emb_dim, hidden_dim, batch_first=True)
        # 最后把 hidden state 转换成词表大小的概率分布
        self.fc_out = nn.Linear(hidden_dim, output_dim)

    def forward(self, input_token, hidden, cell):
        # input_token: [batch_size, 1] (当前这一步输入的词)
        # hidden, cell: 来自 Encoder 或 上一步 Decoder 的状态
        
        # 1. 增加维度: [batch, 1] -> [batch, 1, emb_dim]
        embedded = self.embedding(input_token)
        
        # 2. LSTM 跑一步
        output, (hidden, cell) = self.rnn(embedded, (hidden, cell))
        
        # 3. 预测下一个词: [batch, 1, hidden] -> [batch, 1, vocab_size]
        prediction = self.fc_out(output)
        
        # 去掉序列维度: [batch, vocab_size]
        return prediction.squeeze(1), hidden, cell

class Seq2Seq(nn.Module):
    def __init__(self, encoder, decoder):
        super(Seq2Seq, self).__init__()
        self.encoder = encoder
        self.decoder = decoder

    def forward(self, src, trg):
        # src: 输入序列 [batch, seq_len]
        # trg: 目标序列 [batch, seq_len] (用来训练)
        
        batch_size = src.shape[0]
        trg_len = trg.shape[1]
        trg_vocab_size = VOCAB_SIZE
        
        # 准备一个容器存储 Decoder 每一步的输出
        outputs = torch.zeros(batch_size, trg_len, trg_vocab_size)
        
        # 1. 【Encoder 阶段】：把 source 跑完，拿到 context vector
        hidden, cell = self.encoder(src)
        
        # 2. 【Decoder 阶段】
        # Decoder 的第一个输入是 <SOS> (Start Of Sentence)
        input_token = trg[:, 0].unsqueeze(1) # 取出 SOS
        
        for t in range(1, trg_len):
            # 跑一步 Decoder
            output, hidden, cell = self.decoder(input_token, hidden, cell)
            
            # 存下预测结果
            outputs[:, t, :] = output
            
            # Teacher Forcing: 训练时，下一步的输入直接用正确答案
            # (也可以用 output.argmax(1) 自己预测的，但刚开始训练用正确答案收敛快)
            input_token = trg[:, t].unsqueeze(1)
            
        return outputs

# --- 3. 数据生成器 ---
def get_data():
    # 生成随机序列: [1, 5, 2, 9]
    seq = [random.randint(0, 9) for _ in range(SEQ_LEN)]
    
    # 输入: [1, 5, 2, 9]
    src = torch.LongTensor([seq]) # 加个 batch 维度
    
    # 目标: SOS + [9, 2, 5, 1] + EOS
    target_seq = [SOS_TOKEN] + seq[::-1] + [EOS_TOKEN]
    trg = torch.LongTensor([target_seq])
    
    return src, trg

# --- 4. 训练 ---
# 初始化模型
enc = Encoder(VOCAB_SIZE, EMBED_DIM, HIDDEN_DIM)
dec = Decoder(VOCAB_SIZE, EMBED_DIM, HIDDEN_DIM)
model = Seq2Seq(enc, dec)
optimizer = optim.Adam(model.parameters(), lr=LR)
criterion = nn.CrossEntropyLoss()

print("🚀 开始训练 Seq2Seq (CPU版)...")

for epoch in range(EPOCHS):
    model.train()
    src, trg = get_data()
    
    optimizer.zero_grad()
    
    # 前向传播
    output = model(src, trg)
    
    # 计算 Loss
    # output: [batch, trg_len, vocab_size] -> reshape -> [batch*trg_len, vocab_size]
    # trg: [batch, trg_len] -> reshape -> [batch*trg_len]
    # 忽略第0个(SOS)，从第1个开始算 loss
    loss = criterion(output[:, 1:].reshape(-1, VOCAB_SIZE), trg[:, 1:].reshape(-1))
    
    loss.backward()
    optimizer.step()
    
    if (epoch + 1) % 200 == 0:
        print(f"Epoch {epoch+1}, Loss: {loss.item():.4f}")

print("\n✅ 训练完成！开始测试...")

# --- 5. 测试 (Inference) ---
model.eval()
with torch.no_grad():
    # 生成新数据
    src, trg = get_data()
    print(f"输入序列: {src.squeeze().tolist()}")
    
    # 手动跑一遍 Inference (不使用 Teacher Forcing)
    # 1. Encode
    hidden, cell = model.encoder(src)
    
    # 2. Decode
    input_token = torch.LongTensor([[SOS_TOKEN]]) # 初始输入 SOS
    pred_seq = []
    
    for _ in range(SEQ_LEN + 1): # 最多预测这么长
        output, hidden, cell = model.decoder(input_token, hidden, cell)
        
        # 选概率最大的词
        pred_token = output.argmax(1).item()
        
        if pred_token == EOS_TOKEN:
            break
            
        pred_seq.append(pred_token)
        
        # 把预测的词作为下一步输入
        input_token = torch.LongTensor([[pred_token]])
        
    print(f"模型预测: {pred_seq}")
    print(f"正确答案: {trg.squeeze().tolist()[1:-1]}") # 去掉SOS和EOS看中间